File: 0135-candy/0135-candy.rb
# @param {Integer[]} ratings
# @return {Integer}
def candy(ratings)
  ans = [1]*ratings.length
  
  (0...ratings.length-1).each do |i|
    ans[i+1] = ans[i]+1 if ratings[i+1] > ratings[i]
  end
  
  (1...ratings.length).reverse_each do |i|
    ans[i-1] = [ans[i-1], ans[i]+1].max if ratings[i-1] > ratings[i]
  end
  
  ans.sum
end

File: 0135-candy/0135-candy.go
func candy(ratings []int) int {
  ans := make([]int, len(ratings))
  for i := 0; i < len(ans); i++ {
    ans[i] = 1
  }
  for i := 0; i < len(ratings) - 1; i++ {
    if ratings[i + 1] > ratings[i] {
      ans[i + 1] = ans[i] + 1
    }
  }
  for i := len(ratings) - 1; i >= 1; i-- {
    if ratings[i - 1] > ratings[i] {
      ans[i - 1] = max(ans[i - 1], ans[i] + 1)
    }
  }
  out := 0
  for i := 0; i < len(ans); i++ {
    out += ans[i]
  }
  return out
}